### PR TITLE
fix(pss): make sending PSS messages more efficient

### PR DESCRIPTION
**Pull request by zelig**
a minor improvement on pss trojan mining

- code cleanup
- using errgroup
- change to bigendian 
- fix for the parity bit

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2140)

### CODE REVIEW

**Comment by acud** [comment]
@zelig can you explain why and how this is improving things? (and to which degree?)

**Comment by acud** [comment]
before:
```
growlie@growlie-strikes-back ~/go/src/github.com/ethersphere/bee/pkg/pss ±master⚡ » go test -v -bench=. -run Bench -benchmem -benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/ethersphere/bee/pkg/pss
cpu: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
BenchmarkWrap
BenchmarkWrap/length:1,depth:1
BenchmarkWrap/length:1,depth:1-4         	     187	  33675828 ns/op	18788199 B/op	  104356 allocs/op
BenchmarkWrap/length:4,depth:1
BenchmarkWrap/length:4,depth:1-4         	     622	   9896097 ns/op	 5419940 B/op	   30071 allocs/op
BenchmarkWrap/length:16,depth:1
BenchmarkWrap/length:16,depth:1-4        	    1520	   3850104 ns/op	 2013067 B/op	   11114 allocs/op
BenchmarkWrap/length:16,depth:2
BenchmarkWrap/length:16,depth:2-4        	      30	 757000467 ns/op	425943036 B/op	 2366465 allocs/op
BenchmarkWrap/length:64,depth:2
BenchmarkWrap/length:64,depth:2-4        	      38	 145133468 ns/op	81234541 B/op	  451305 allocs/op
BenchmarkWrap/length:256,depth:2
BenchmarkWrap/length:256,depth:2-4       	     166	  37935671 ns/op	21093620 B/op	  117162 allocs/op
BenchmarkWrap/length:256,depth:3
BenchmarkWrap/length:256,depth:3-4       	       1	6143139754 ns/op	3259615688 B/op	18111114 allocs/op
BenchmarkWrap/length:4096,depth:3
BenchmarkWrap/length:4096,depth:3-4      	      15	 591351344 ns/op	242435498 B/op	 1347379 allocs/op
BenchmarkWrap/length:16384,depth:3
BenchmarkWrap/length:16384,depth:3-4     	      40	 223228374 ns/op	74415670 B/op	  413806 allocs/op
PASS
ok  	github.com/ethersphere/bee/pkg/pss	106.122s
```

after:
```
growlie@growlie-strikes-back ~/go/src/github.com/ethersphere/bee/pkg/pss ±(dfccf46b...)⚡ » go test -v -bench=. -run Bench -benchmem -benchtime 5s
goos: linux
goarch: amd64
pkg: github.com/ethersphere/bee/pkg/pss
cpu: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
BenchmarkWrap
BenchmarkWrap/length:1,depth:1
BenchmarkWrap/length:1,depth:1-4         	     232	  27393762 ns/op	19365715 B/op	  107572 allocs/op
BenchmarkWrap/length:4,depth:1
BenchmarkWrap/length:4,depth:1-4         	     844	   7616651 ns/op	 5263813 B/op	   29211 allocs/op
BenchmarkWrap/length:16,depth:1
BenchmarkWrap/length:16,depth:1-4        	    1869	   3619230 ns/op	 1997974 B/op	   11042 allocs/op
BenchmarkWrap/length:16,depth:2
BenchmarkWrap/length:16,depth:2-4        	      13	 435784661 ns/op	261811824 B/op	 1454861 allocs/op
BenchmarkWrap/length:64,depth:2
BenchmarkWrap/length:64,depth:2-4        	      61	 113410167 ns/op	75256698 B/op	  418116 allocs/op
BenchmarkWrap/length:256,depth:2
BenchmarkWrap/length:256,depth:2-4       	     206	  30770748 ns/op	18623767 B/op	  103452 allocs/op
BenchmarkWrap/length:256,depth:3
BenchmarkWrap/length:256,depth:3-4       	       1	9820205532 ns/op	3980843456 B/op	22129769 allocs/op
BenchmarkWrap/length:4096,depth:3
BenchmarkWrap/length:4096,depth:3-4      	       9	 956932108 ns/op	326600746 B/op	 1815988 allocs/op
BenchmarkWrap/length:16384,depth:3
BenchmarkWrap/length:16384,depth:3-4     	      26	 264735541 ns/op	85486048 B/op	  475630 allocs/op
PASS
ok  	github.com/ethersphere/bee/pkg/pss	80.735s
```

File: pkg/pss/trojan.go [hunk 1]
```diff
@@ -214,51 +215,48 @@ func mine(ctx context.Context, odd bool, f func(nonce []byte) (swarm.Chunk, erro
...
 				}
 				if res != nil {
 					result <- res
-					return
-				}
-				select {
-				case <-quit:
-					return
-				default:
+					return errors.New("found")
```
Comment (acud) [inline pkg/pss/trojan.go:248]: why is it needed to return an error when there was no error and a result was communicated?

**Review by acud — CHANGES_REQUESTED** [review]

**Review by mrekucci — COMMENTED** [review]
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2140)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @acud and @aloknerurkar)

___
*[pkg/pss/trojan.go, line 248 at r1](https://reviewable.io/reviews/ethersphere/bee/2140#-McYRWnsCVYTPvnuwnKi-r1-248:-Md1JFnA7hOcnsAl5HWr:b-s22gdm) ([raw file](https://github.com/ethersphere/bee/blob/dfccf46bdaed159ef114f7da53fcbcd9964a3cef/pkg/pss/trojan.go#L248)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

why is it needed to return an error when there was no error and a result was communicated?
</blockquote></details>

Agree with @acud here. You may return `nil` here (or more appropriate, `break` the loop and return `nil` at the end of the gorutine closure) and then simplify the section with `eg.Wait()` at the end as:
```golang
if err := eg.Wait(); err != nil {
	return nil, err
}
return <-result, nil
```

File: pkg/pss/trojan.go (hunk 1 above)
Comment (zelig) [reply pkg/pss/trojan.go:248]: this was only to trigger the context cancel to terminate the other routines, but i guess i should just use a cancel on the context right?

**Comment by aloknerurkar** [comment]
___
*[pkg/pss/trojan.go, line 248 at r1](https://reviewable.io/reviews/ethersphere/bee/2140#-McYRWnsCVYTPvnuwnKi-r1-248:-Md1fe9zAgxeGPVrZQFP:b5dpars) ([raw file](https://github.com/ethersphere/bee/blob/dfccf46bdaed159ef114f7da53fcbcd9964a3cef/pkg/pss/trojan.go#L248)):*
<details><summary><i>Previously, zelig (Viktor Trón) wrote…</i></summary><blockquote>

this was only to trigger the context cancel to terminate the other routines, but i guess i should just use a cancel on the context right?
</blockquote></details>

Context cancellation will return error on eg.Wait from the next routine. If we dont care about the errors from the other routines once the first one is done, we should `return nil` on `<-ctx.Done()`. Although I dont think errorgroup usage give any advantages here apart from context semantics

**Comment by mrekucci** [comment]
> _[pkg/pss/trojan.go, line 248 at r1](https://reviewable.io/reviews/ethersphere/bee/2140#-McYRWnsCVYTPvnuwnKi-r1-248:-Md1fe9zAgxeGPVrZQFP:b5dpars) ([raw file](https://github.com/ethersphere/bee/blob/dfccf46bdaed159ef114f7da53fcbcd9964a3cef/pkg/pss/trojan.go#L248)):_
> Previously, zelig (Viktor Trón) wrote…
> 
> Context cancellation will return error on eg.Wait from the next routine. If we dont care about the errors from the other routines once the first one is done, we should `return nil` on `<-ctx.Done()`. Although I dont think errorgroup usage give any advantages here apart from context semantics

Agree, we can also specifically check if the returned error is the `context.Canceled` error and ignore it in the it is.

File: pkg/pss/trojan.go [hunk 2]
```diff
@@ -203,62 +201,54 @@ func contains(col Targets, elem []byte) bool {
...
-	case res := <-result:
-		return res, nil
+		})
 	}
+	go func() {
+		err = eg.Wait()
+		result <- nil
+	}()
+	// if there was an error it gets set
+	return <-result, err
```
Comment (aloknerurkar) [inline pkg/pss/trojan.go:251]: I am still not 100% sure if we cover all the cases here. I would feel more comfortable if we could test some of these things in the unit tests. For eg. delaying some of the function calls passed as arguments here.

File: pkg/pss/trojan.go (hunk 1 above)
Comment (acud) [reply pkg/pss/trojan.go]: yes :+1:

**Review by acud — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2140#-:-MdVjxyO9vUbAMzRoYf5:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @aloknerurkar, @mrekucci, and @zelig)

**Review by mrekucci — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2140)* status: all files reviewed, 2 unresolved discussions (waiting on @aloknerurkar, @mrekucci, and @zelig)

**Comment by acud** [comment]
@zelig pss integration tests are failing

**Comment by acud** [comment]
seems to be OK for now. restarted the build several times. let's monitor the situation

**Pull request merged**